src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java

```diff
@@ -75,20 +80,44 @@
         return bindings;
     }
 
-    public static final class Execution extends AbstractStepExecutionImpl {
+    @Override
+    public StepExecution start(StepContext context) throws Exception {
+        return new Execution(this, context);
+    }
+
+    public static final class Execution extends StepExecution {
 
         private static final long serialVersionUID = 1;
 
-        @Inject(optional=true) private transient BindingStep step;
-        @StepContextParameter private transient Run<?,?> run;
-        @StepContextParameter private transient FilePath workspace;
-        @StepContextParameter private transient Launcher launcher;
-        @StepContextParameter private transient TaskListener listener;
+        private transient BindingStep step;
+
+        public Execution(@Nonnull BindingStep step, StepContext context) {
+            super(context);
+            this.step = step;
+        }
 
         @Override public boolean start() throws Exception {
+            Run<?,?> run = getContext().get(Run.class);
+            if (run == null) {
+                throw new MissingContextVariableException(Run.class);
+            }
+            TaskListener listener = getContext().get(TaskListener.class);
+            if (listener == null) {
+                throw new MissingContextVariableException(TaskListener.class);
+            }
+
+            FilePath workspace = getContext().get(FilePath.class);
+            Launcher launcher = getContext().get(Launcher.class);
+
             Map<String,String> overrides = new HashMap<String,String>();
             List<MultiBinding.Unbinder> unbinders = new ArrayList<MultiBinding.Unbinder>();
             for (MultiBinding<?> binding : step.bindings) {
+                for (Class<?> requiredContext : binding.getDescriptor().getRequiredContext()) {
+                    Object v = getContext().get(requiredContext);
+                    if (v == null) {
+                        throw new MissingContextVariableException(requiredContext);
+                    }
+                }
                 MultiBinding.MultiEnvironment environment = binding.bind(run, workspace, launcher, listener);
```

Comment:
Hmm. So `MultiBinding.bind`’s signature should be changed to make `workspace` and `launcher` `@Nullable`, with Javadoc explaining whether the implementation can or cannot rely on them being present. Same for `Unbinder`. (`listener` should be marked `@Nonnull` like `build` already was.)